### PR TITLE
[FIX] mail: allow tests to await specific events

### PR DESCRIPTION
**Pull request by seb-odoo**

### CODE REVIEW

**Comment by seb-odoo** [comment]
POC done, the current syntax is too verbose but it does the job.

File: addons/mail/static/src/components/discuss/tests/discuss_tests.js [hunk 1]
```diff
@@ -1875,34 +1865,42 @@ QUnit.test('restore thread scroll position', async function (assert) {
         "should have 25 messages in channel 11"
     );
 
-    // scroll to top of channel11
-    await afterNextRender(() => {
-        document.querySelector(`.o_Discuss_thread .o_ThreadView_messageList`).scrollTop = 0;
+    await this.afterEvent({
+        eventName: 'o-component-message-list-scrolled',
+        func: () => document.querySelector(`.o_Discuss_thread .o_ThreadView_messageList`).scrollTop = 0,
+        message: 'should wait until channel 11 changed its scroll position to top',
```
Comment (alexkuhn) [inline addons/mail/static/src/components/discuss/tests/discuss_tests.js:1871]: double quotes

File: addons/mail/static/src/components/discuss/tests/discuss_tests.js [hunk 2]
```diff
@@ -1875,34 +1865,42 @@ QUnit.test('restore thread scroll position', async function (assert) {
...
+                .o_DiscussSidebar_groupChannel
+                .o_DiscussSidebar_item[data-thread-local-id="${
+                    this.env.models['mail.thread'].find(thread =>
+                        thread.id === 12 &&
+                        thread.model === 'mail.channel'
+                    ).localId
+                }"]
+            `).click();
+        },
+        message: 'should wait until channel 12 scrolled to its last message',
```
Comment (alexkuhn) [inline addons/mail/static/src/components/discuss/tests/discuss_tests.js:1900]: double quotes

File: addons/mail/static/src/components/discuss/tests/discuss_tests.js [hunk 3]
```diff
@@ -1912,43 +1910,50 @@ QUnit.test('restore thread scroll position', async function (assert) {
...
+                .o_DiscussSidebar_groupChannel
+                .o_DiscussSidebar_item[data-thread-local-id="${
+                    this.env.models['mail.thread'].find(thread =>
+                        thread.id === 11 &&
+                        thread.model === 'mail.channel'
+                    ).localId
+                }"]
+            `).click();
+        },
+        message: 'should wait until channel 11 restored its scroll position',
```
Comment (alexkuhn) [inline addons/mail/static/src/components/discuss/tests/discuss_tests.js:1927]: double quotes

File: addons/mail/static/src/components/discuss/tests/discuss_tests.js [hunk 4]
```diff
@@ -1912,43 +1910,50 @@ QUnit.test('restore thread scroll position', async function (assert) {
...
+                .o_DiscussSidebar_groupChannel
+                .o_DiscussSidebar_item[data-thread-local-id="${
+                    this.env.models['mail.thread'].find(thread =>
+                        thread.id === 12 &&
+                        thread.model === 'mail.channel'
+                    ).localId
+                }"]
+            `).click();
+        },
+        message: 'should wait until channel 12 recovered its scroll position',
```
Comment (alexkuhn) [inline addons/mail/static/src/components/discuss/tests/discuss_tests.js:1952]: double quotes

**Review by alexkuhn — APPROVED** [review]
Definitely more expressive than `afterNextRender`.
Sorry for being rude, but it's not great to read and use.
If we take the time and effort to design a logging system, I'm sure we could have a much nicer system that could fix this issue.

I don't mind merging this solution in the meantime, because it fixes a race condition and it's more important to fix this than fixing the unpleasantness.

**Review by alexkuhn — APPROVED** [review]
@robodoo r+

**Pull request closed without merging**